### PR TITLE
End epoch feature

### LINKED ISSUE #1882: Add "End Epoch" button and functionality.
Add a "End Epoch" button that ends the current epoch and sets the end date to right now. This is only shown on the 

Image

This will pop a modal to verify and confirm ending.

Figma:

https://www.figma.com/file/L4oZwYVUdpgacZtwipGaYe/Coordinape-App?node-id=31562%3A89903&t=EyZz5N2IsSH3dmfZ-0

This work should be based off the latest Epoch UI branch: https://github.com/coordinape/coordinape/tree/epoch-form-refinements

### PR DESCRIPTION
**Pull request by sshmm**
# MERGE PLAN:

This can be deployed after the new epoch stuff is deployed. It is not a blocker for the others to be merged.

## Motivation and Context

resolved #1882

## Description

add end epoch button to current epoch

## Test and Deployment Plan
preconditon: active epoch with gifts
1- Go to hisory page and click edit epoch on current epoch
2- click end epoch and confirm on the warning modal
3- the epoch should end and the gifts are refilled

## Screenshots (if appropriate):

image
image

### CODE REVIEW

File: src/pages/HistoryPage/EndEpochDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  const queryClient = useQueryClient();
+  const { showError } = useToast();
+
+  const endEpoch = useMutation(endEpochMutation, {
+    onMutate: () => {
+      setSubmitting(true);
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries(QUERY_KEY_ACTIVE_HISTORY);
```
Comment (crabsinger) [inline src/pages/HistoryPage/EndEpochDialog.tsx:35]: Let's add a showSuccess toast from useToast here on success as well as error.

**Comment by sshmm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1771** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1771" target="_blank">Graphite</a> 
    * **PR #1772** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1772" target="_blank">Graphite</a> 
      * **PR #1773** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1773" target="_blank">Graphite</a> 
        * **PR #1829** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1829" target="_blank">Graphite</a> 
          * **PR #1832** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1832" target="_blank">Graphite</a> 
            * **PR #1774** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1774" target="_blank">Graphite</a> 
              * **PR #1903** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1903" target="_blank">Graphite</a> 
                * **PR #1901** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1901" target="_blank">Graphite</a>  👈
            * **PR #1866** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1866" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1901?utm_source=stack-comment).

**Comment by sshmm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1771** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1771" target="_blank">Graphite</a> 
    * **PR #1772** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1772" target="_blank">Graphite</a> 
      * **PR #1773** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1773" target="_blank">Graphite</a> 
        * **PR #1829** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1829" target="_blank">Graphite</a> 
          * **PR #1832** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1832" target="_blank">Graphite</a> 
            * **PR #1774** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1774" target="_blank">Graphite</a> 
              * **PR #1903** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1903" target="_blank">Graphite</a> 
                * **PR #1901** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1901" target="_blank">Graphite</a>  👈
            * **PR #1866** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1866" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1901?utm_source=stack-comment).

File: src/pages/HistoryPage/EndEpochDialog.tsx (hunk 1 above)
Comment (crabsinger) [reply src/pages/HistoryPage/EndEpochDialog.tsx:35]: @sshmm

File: api/hasura/actions/_handlers/endEpoch.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  id: z.number(),
+  circle_id: z.number(),
+});
+
+async function handler(request: VercelRequest, response: VercelResponse) {
+  const {
+    input: { payload: input },
+  } = composeHasuraActionRequestBody(EpochEndSchema).parse(request.body);
+
+  const deletingEpoch = await getExistingEpoch(input);
```
Comment (crabsinger) [inline api/hasura/actions/_handlers/endEpoch.ts:23]: @sshmm this is just being ended, not deleted, right?
```suggestion
  const endingEpoch = await getExistingEpoch(input);
```

File: src/pages/HistoryPage/EndEpochDialog.tsx (hunk 1 above)
Comment (sshmm) [reply src/pages/HistoryPage/EndEpochDialog.tsx:35]: done
image

**Review by crabsinger — APPROVED** [review]

**Pull request merged**